### DISCUSSION TITLE
How to use generated "Route<T>" type (from "typedRoutes") with dynamic routes

### DISCUSSION THREAD

**Discussion by borispetrovdev** [vercel/next.js#69018] [App Router]
TL;DR: In the following example code, why am I getting the error `"/dynamic/foo"' is not assignable to type 'Route'`, despite the existence of the route `/dynamic/[some_param]`?
![image](https://github.com/user-attachments/assets/22885b65-a2b3-429e-875f-a04f1cb47f33)

Details: I'm trying to achieve type-safe `redirect` (and `revalidatePath`) calls using the experimental `typedRoutes` feature. Since `typedRoutes` doesn't automatically ensure that strings passed to `redirect` correspond to existing routes (it seems to only type-check the `href` prop of `<Link>`), I am declaring a `typedRoute: Route` variable and then passing it to `redirect`. However, I'm getting a typescript error essentially saying that the string I've assigned to `typedRoute` is not a valid `Route`. 

As you can see in the folder structure pictured in the screenshot above, I have a route `/dynamic/[some_param]`, so the string `"/dynamic/foo"` should indeed be a valid `Route`. In fact, when I inspect the `Route<T>` type, I see `/dynamic/$__next_route_internal_types_SafeSlug<infer _ extends string>}`, so the route _is_ correctly detected by `typedRoutes`, but I can't seem to access it when I'm assigning types (and values) to variables:
![image](https://github.com/user-attachments/assets/d89f68c9-ef8b-478b-af2d-2a27d1c9cf80)

Ideally, I would write `const typedRoute: Route = "/dyna"` and get an autocomplete saying `/dynamic/${string}`. What am I missing?

**Comment by controversial** [comment]
You have to also provide the string as a parameter to the `Route` type:
```ts
const typedRoute: Route<'/dynamic/foo'> = '/dynamic/foo';
```
when the Route type is used in context (e.g. in  the`<Link>` component) the type parameter is usually inferred, however, if you’re referring to the `Route` type explicitly, you need to make sure the appropriate type parameter is passed, rather than the default `Route<string>`.

Also, I’d recommend using `satisfies` over type annotations in this case, in order to preserve a more specific type:
```ts
const typedRoute = '/dynamic/foo' satisfies Route<'/dynamic/foo'>;
```

---

If you want to simplify the syntax so that the type parameter can be inferred in context, you could make a function that infers the type parameter:
```ts
function typedRoute<T extends string>(route: Route<T>) {
  return route;
}

const myRoute = typedRoute('/dynamic/foo');
// myRoute has type Route<'/dynamic/foo'>, and will error if an invalid route is passed
```

**Reply by uncvrd** [reply]
thanks for this solution, `typedRoute` works! kinda stinks that you dont get autocomplete for dynamic routes but I guess it makes sense...

**Comment by toppsdown** [comment]
Typed routes is working fine for static, but I cannot for the life of me get it to work with dynamic routes.

The provided example code gives me no type error:
```typescript
import { Route } from "next"

function typedRoute<T extends string>(route: T): Route<T> {
  return route as Route<T>
}

const _myRoute = typedRoute("/not a real route")

```

Providing the path to the Route type sorta works, but it doesn't detect if the route changes:

```typescript
function getEventUrl(event: Event): Route<`/events/${string}/${string}`> {
  const eventName = event.name.toLowerCase().replace(/[^a-z0-9]+/g, "-")

  const eventId = event.id
  const eventNameSlug = eventName

  return `/something else/${eventId}/${eventNameSlug}` // This properly errors
}
```

```typescript
function getEventUrl(event: Event): Route<`/not-a-path/${string}/${string}`> { // this does not error
  const eventName = event.name.toLowerCase().replace(/[^a-z0-9]+/g, "-")

  const eventId = event.id
  const eventNameSlug = eventName

  return `/not-a-path/${eventId}/${eventNameSlug}` 
}
```

It seems that the Route export doesn't actually export the DynamicTypes, which I don't understand. It only exports the type matcher.

```typescript
type RouteImpl<T> = 
    | StaticRoutes
    | SearchOrHash
    | WithProtocol
    | `${StaticRoutes}${SearchOrHash}`
    | (T extends `${DynamicRoutes<infer _>}${Suffix}` ? T : never) // type matcher
```

If I add `| DynamicRoutes<T>` to the route implementation, then all the types seem to work properly.

**Reply by ealexhaywood** [reply]
Yep I'm running into this same issue getting dynamic routes to work.  I see it correctly generates all of them as the `DynamicRoutes` type but no dice on getting them to work with any functions that produce dynamic URLs, like your `getEventUrl()`

I'm doing something like this

```ts
function MyComponent() {
  const { id } = useParams<{ id: string }>();
  const navItems = [
   {
      title: "Section 1",
      href: `/dynamic/route/${id}`,
    },
    {
      title: "Section 2",
      href: `/dynamic/route/${id}/subpage`
    }
  ] satisfies { title: string; href: Route } // this is the same type expected by SidebarNav
  
  return <SidebarNav items={items} {...otherProps} />
}
```

I've tried a bunch of things and nothing actually works outside of just lying to the TS compiler.

**Reply by toppsdown** [reply]
I ended up patching next.js typed routes and it's been working ever since. Honestly forgot about the issue.

```
diff --git a/node_modules/next/dist/build/webpack/plugins/next-types-plugin/index.js b/node_modules/next/dist/build/webpack/plugins/next-types-plugin/index.js
index 30e27e2..f28c797 100644
--- a/node_modules/next/dist/build/webpack/plugins/next-types-plugin/index.js
+++ b/node_modules/next/dist/build/webpack/plugins/next-types-plugin/index.js
@@ -364,6 +364,8 @@ declare namespace __next_route_internal_types__ {
     | WithProtocol
     | \`\${StaticRoutes}\${SearchOrHash}\`
     | (T extends \`\${DynamicRoutes<infer _>}\${Suffix}\` ? T : never)
+    | DynamicRoutes<T>
+    | \`\${DynamicRoutes<T>}\${Suffix}\`
     `}
 }
 
```

It's not 100% correct since it matches everything after the first slug, but it's good enough for my needs
